File: worker.go
package main

import (
	"fmt"
	"log/slog"
	"sync"
	"time"
)

type Worker struct {
	cfg      *Config
	svc      *Service
	msgStore *MessageStore
	stopCh   chan struct{}
	wg       sync.WaitGroup
	mu       sync.Mutex
	started  bool
}

func NewWorker(cfg *Config, svc *Service, msgStore *MessageStore) *Worker {
	return &Worker{
		cfg:      cfg,
		svc:      svc,
		msgStore: msgStore,
		stopCh:   make(chan struct{}),
	}
}

func (w *Worker) Start() {
	w.mu.Lock()
	defer w.mu.Unlock()

	if w.started {
		return
	}
	w.started = true

	w.wg.Add(3)
	go w.runTableManager()
	go w.runMaintenance()
	go w.runIdempotencyCleanup()
	slog.Info("Worker started")
}

func (w *Worker) Stop() {
	w.mu.Lock()
	if !w.started {
		w.mu.Unlock()
		return
	}
	w.started = false
	w.mu.Unlock()

	close(w.stopCh)
	w.wg.Wait()
	slog.Info("Worker stopped")
}

// runLoop 通用轮询循环：等待初始延迟 → 定时执行 fn
func (w *Worker) runLoop(delay time.Duration, interval time.Duration, fn func()) {
	defer w.wg.Done()

	select {
	case <-time.After(delay):
	case <-w.stopCh:
		return
	}

	ticker := time.NewTicker(interval)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			fn()
		case <-w.stopCh:
			return
		}
	}
}

// ==================== TableManager ====================

func (w *Worker) runTableManager() {
	w.runLoop(calculateNextMondayDelay(), time.Duration(w.cfg.WorkerTableCreateIntervalHours)*time.Hour, w.createWeekTables)
}

func (w *Worker) createTablesFromTodayToSunday() error {
	now := time.Now().UTC()
	dayOfWeek := int(now.Weekday())
	daysToSunday := 0
	if dayOfWeek != 0 {
		daysToSunday = 7 - dayOfWeek
	}

	var lastErr error
	for offset := 0; offset <= daysToSunday; offset++ {
		date := now.AddDate(0, 0, offset)
		tableName := MessageTableNameByDate(date)
		if err := w.msgStore.EnsurePartition(tableName); err != nil {
			slog.Error("failed to create table", "table", tableName, "error", err)
			lastErr = err
		}
	}
	slog.Info("initial message tables created")
	return lastErr
}

// CreateTablesFromTodayToSundayWithError 带错误返回的版本，用于启动时检查
func (w *Worker) CreateTablesFromTodayToSundayWithError() error {
	return w.createTablesFromTodayToSunday()
}

func (w *Worker) createWeekTables() {
	now := time.Now().UTC()
	weekday := int(now.Weekday())
	if weekday == 0 {
		weekday = 7
	}
	nextMonday := now.AddDate(0, 0, 8-weekday)

	for offset := 0; offset < 7; offset++ {
		date := nextMonday.AddDate(0, 0, offset)
		tableName := MessageTableNameByDate(date)
		if err := w.msgStore.EnsurePartition(tableName); err != nil {
			slog.Error("failed to create table", "table", tableName, "error", err)
		}
	}
	slog.Info("weekly message tables created")
}

func calculateNextMondayDelay() time.Duration {
	now := time.Now().UTC()
	daysUntilMonday := (8 - int(now.Weekday())) % 7
	if daysUntilMonday == 0 {
		daysUntilMonday = 7
	}
	nextMonday := time.Date(now.Year(), now.Month(), now.Day()+daysUntilMonday,
		0, 0, 0, 0, time.UTC)
	return nextMonday.Sub(now)
}

// ==================== Maintenance ====================

func (w *Worker) runMaintenance() {
	w.runLoop(time.Duration(w.cfg.WorkerMaintenanceInitialDelayMinutes)*time.Minute,
		time.Duration(w.cfg.WorkerMaintenanceIntervalHours)*time.Hour,
		func() { w.runAnalyze(); w.dropExpiredPartitions() })
}

func (w *Worker) runAnalyze() {
	_, err := w.msgStore.DB().Exec("ANALYZE messages")
	if err != nil {
		slog.Error("maintenance: ANALYZE failed", "error", err)
		return
	}
	slog.Info("maintenance: ANALYZE completed")
}

func (w *Worker) dropExpiredPartitions() {
	retentionDays := w.cfg.MessageRetentionDays
	cutoff := time.Now().UTC().AddDate(0, 0, -retentionDays)

	rows, err := w.msgStore.DB().Query(`
		SELECT inhrelid::regclass::text
		FROM pg_inherits
		WHERE inhparent = 'messages'::regclass
	`)
	if err != nil {
		slog.Error("maintenance: failed to list partitions", "error", err)
		return
	}
	defer rows.Close()

	var partitions []string
	for rows.Next() {
		var name string
		if err := rows.Scan(&name); err != nil {
			slog.Error("maintenance: failed to scan partition name", "error", err)
			continue
		}
		partitions = append(partitions, name)
	}

	for _, partition := range partitions {
		if len(partition) < 10 || partition[:9] != "messages_" {
			continue
		}
		dateStr := partition[9:]
		partitionDate, err := time.Parse("20060102", dateStr)
		if err != nil {
			continue
		}
		if partitionDate.Before(cutoff) {
			var quoted string
			err := w.msgStore.DB().QueryRow("SELECT quote_ident($1)", partition).Scan(&quoted)
			if err != nil {
				slog.Error("maintenance: failed to quote partition name", "partition", partition, "error", err)
				continue
			}
			_, err = w.msgStore.DB().Exec(fmt.Sprintf("DROP TABLE IF EXISTS %s", quoted))
			if err != nil {
				slog.Error("maintenance: failed to drop partition", "partition", partition, "error", err)
				continue
			}
			slog.Info("maintenance: dropped partition", "partition", partition)
		}
	}
	slog.Info("maintenance: drop expired partitions completed")
}

func MessageTableNameByDate(date time.Time) string {
	return fmt.Sprintf("messages_%s", date.Format("20060102"))
}

func MessageTableNameByTs(ts int64) string {
	t := time.UnixMilli(ts)
	return MessageTableNameByDate(t)
}

// ==================== Idempotency Cleanup ====================

func (w *Worker) runIdempotencyCleanup() {
	w.runLoop(time.Duration(w.cfg.WorkerMaintenanceInitialDelayMinutes)*time.Minute,
		time.Duration(w.cfg.WorkerMaintenanceIntervalHours)*time.Hour,
		w.cleanupIdempotency)
}

func (w *Worker) cleanupIdempotency() {
	retentionHours := int64(24) // 默认保留24小时
	cutoff := time.Now().UnixMilli() - retentionHours*3600*1000
	result, err := w.msgStore.DB().Exec(`DELETE FROM message_idempotency WHERE created_at < $1`, cutoff)
	if err != nil {
		slog.Error("idempotency cleanup failed", "error", err)
		return
	}
	count, _ := result.RowsAffected()
	if count > 0 {
		slog.Info("cleaned up stale idempotency records", "count", count)
	}
}

File: src/worker.rs
use crate::config::Config;
use crate::store::MessageStore;
use chrono::{Datelike, Duration, NaiveDate, Timelike, Utc};
use sqlx::PgPool;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use tokio::sync::Mutex;
use tokio::time::{interval, sleep, Duration as TokioDuration};

pub struct Worker {
    cfg: Config,
    pool: PgPool,
    msg_store: MessageStore,
    stop_flag: Arc<AtomicBool>,
    started: Arc<Mutex<bool>>,
}

impl Worker {
    pub fn new(cfg: Config, pool: PgPool, msg_store: MessageStore) -> Self {
        Self {
            cfg,
            pool,
            msg_store,
            stop_flag: Arc::new(AtomicBool::new(false)),
            started: Arc::new(Mutex::new(false)),
        }
    }

    pub async fn start(&self) {
        let mut started = self.started.lock().await;
        if *started {
            return;
        }
        *started = true;
        drop(started);

        let stop_flag = self.stop_flag.clone();
        let pool = self.pool.clone();
        let msg_store = self.msg_store.clone();
        let cfg = self.cfg.clone();

        tokio::spawn(async move {
            run_table_manager(stop_flag.clone(), pool.clone(), msg_store.clone(), cfg.clone()).await;
        });

        let stop_flag = self.stop_flag.clone();
        let pool = self.pool.clone();
        let msg_store = self.msg_store.clone();
        let cfg = self.cfg.clone();

        tokio::spawn(async move {
            run_maintenance(stop_flag.clone(), pool.clone(), msg_store.clone(), cfg.clone()).await;
        });

        tracing::info!("Worker started");
    }

    pub async fn stop(&self) {
        let mut started = self.started.lock().await;
        if !*started {
            return;
        }
        *started = false;
        drop(started);

        self.stop_flag.store(true, Ordering::SeqCst);
        tracing::info!("Worker stopped");
    }

    pub fn create_tables_from_today_to_sunday(&self) {
        let now = Utc::now();
        let day_of_week = now.weekday().num_days_from_sunday();
        let days_to_sunday = if day_of_week == 0 { 0 } else { 7 - day_of_week as i32 };

        for offset in 0..=days_to_sunday {
            let date = now + Duration::days(offset as i64);
            let table_name = message_table_name_by_date(date);
            if let Err(e) = self.msg_store.ensure_partition(&table_name) {
                tracing::error!("failed to create table: table={} error={}", table_name, e);
            }
        }
        tracing::info!("initial message tables created");
    }

    pub fn create_tables_from_today_to_sunday_with_error(&self) -> Result<(), String> {
        let now = Utc::now();
        let day_of_week = now.weekday().num_days_from_sunday();
        let days_to_sunday = if day_of_week == 0 { 0 } else { 7 - day_of_week as i32 };

        let mut last_err = None;
        for offset in 0..=days_to_sunday {
            let date = now + Duration::days(offset as i64);
            let table_name = message_table_name_by_date(date);
            if let Err(e) = self.msg_store.ensure_partition(&table_name) {
                tracing::error!("failed to create table: table={} error={}", table_name, e);
                last_err = Some(e);
            }
        }
        tracing::info!("initial message tables created");

        if let Some(e) = last_err {
            Err(e)
        } else {
            Ok(())
        }
    }
}

async fn run_table_manager(
    stop_flag: Arc<AtomicBool>,
    pool: PgPool,
    msg_store: MessageStore,
    cfg: Config,
) {
    let next_monday = calculate_next_monday_delay();
    let delay = TokioDuration::from_secs(next_monday.as_secs());

    tokio::select! {
        _ = sleep(delay) => {}
        _ = async {
            while !stop_flag.load(Ordering::SeqCst) {
                tokio::time::sleep(TokioDuration::from_millis(100)).await;
            }
        } => return,
    }

    let interval_hours = cfg.worker_table_create_interval_hours;
    let mut ticker = interval(TokioDuration::from_secs(interval_hours * 3600));

    loop {
        ticker.tick().await;
        if stop_flag.load(Ordering::SeqCst) {
            return;
        }
        create_week_tables(&pool, &msg_store).await;
    }
}

async fn run_maintenance(
    stop_flag: Arc<AtomicBool>,
    pool: PgPool,
    msg_store: MessageStore,
    cfg: Config,
) {
    let initial_delay = cfg.worker_maintenance_initial_delay_minutes;
    let delay = TokioDuration::from_secs(initial_delay * 60);

    tokio::select! {
        _ = sleep(delay) => {}
        _ = async {
            while !stop_flag.load(Ordering::SeqCst) {
                tokio::time::sleep(TokioDuration::from_millis(100)).await;
            }
        } => return,
    }

    let interval_hours = cfg.worker_maintenance_interval_hours;
    let mut ticker = interval(TokioDuration::from_secs(interval_hours * 3600));

    loop {
        ticker.tick().await;
        if stop_flag.load(Ordering::SeqCst) {
            return;
        }
        run_analyze(&pool).await;
        drop_expired_partitions(&pool, &msg_store, cfg.message_retention_days).await;
    }
}

async fn run_analyze(pool: &PgPool) {
    match sqlx::query("ANALYZE messages").execute(pool).await {
        Ok(_) => tracing::info!("maintenance: ANALYZE completed"),
        Err(e) => tracing::error!("maintenance: ANALYZE failed: error={}", e),
    }
}

async fn drop_expired_partitions(pool: &PgPool, _msg_store: &MessageStore, retention_days: i32) {
    let cutoff = Utc::now() - Duration::days(retention_days as i64);

    let rows = match sqlx::query_scalar::<_, String>(
        r#"
        SELECT inhrelid::regclass::text
        FROM pg_inherits
        WHERE inhparent = 'messages'::regclass
        "#,
    )
    .fetch_all(pool)
    .await
    {
        Ok(rows) => rows,
        Err(e) => {
            tracing::error!("maintenance: failed to list partitions: error={}", e);
            return;
        }
    };

    for partition in rows {
        if partition.len() < 10 || !partition.starts_with("messages_") {
            continue;
        }
        let date_str = &partition[9..];
        let partition_date = match NaiveDate::parse_from_str(date_str, "%Y%m%d") {
            Ok(d) => d,
            Err(_) => continue,
        };

        let partition_utc = partition_date.and_hms_opt(0, 0, 0).unwrap().and_utc();

        if partition_utc < cutoff {
            let quoted = format!("\"{}\"", partition.replace('"', "\"\""));
            let query = format!("DROP TABLE IF EXISTS {}", quoted);
            match sqlx::query(&query).execute(pool).await {
                Ok(_) => tracing::info!("maintenance: dropped partition: partition={}", partition),
                Err(e) => tracing::error!(
                    "maintenance: failed to drop partition: partition={} error={}",
                    partition,
                    e
                ),
            }
        }
    }
    tracing::info!("maintenance: drop expired partitions completed");
}

async fn create_week_tables(pool: &PgPool, msg_store: &MessageStore) {
    let now = Utc::now();
    let weekday = now.weekday().num_days_from_sunday();
    let weekday = if weekday == 0 { 7 } else { weekday as i32 };
    let next_monday = now + Duration::days((8 - weekday) as i64);

    for offset in 0..7 {
        let date = next_monday + Duration::days(offset as i64);
        let table_name = message_table_name_by_date(date);
        if let Err(e) = msg_store.ensure_partition(&table_name) {
            tracing::error!("failed to create table: table={} error={}", table_name, e);
        }
    }
    tracing::info!("weekly message tables created");
}

fn calculate_next_monday_delay() -> Duration {
    let now = Utc::now();
    let days_until_monday = (8 - now.weekday().num_days_from_sunday() as i32) % 7;
    let days_until_monday = if days_until_monday == 0 { 7 } else { days_until_monday };
    let next_monday = now.date_naive() + Duration::days(days_until_monday as i64);
    let next_monday_utc = next_monday.and_hms_opt(0, 0, 0).unwrap().and_utc();
    next_monday_utc - now
}

pub fn message_table_name_by_date(date: chrono::DateTime<Utc>) -> String {
    format!("messages_{}", date.format("%Y%m%d"))
}

pub fn message_table_name_by_ts(ts: i64) -> String {
    let dt = chrono::DateTime::<Utc>::from_timestamp_millis(ts).unwrap_or_default();
    message_table_name_by_date(dt)
}
